### PR TITLE
[METEOR-1208][REFACTOR] Allow the user to select whether to create or load a project on init

### PR DESCRIPTION
**Pull request by patrickcleeve2**
Currently, when starting the odemis-gui for meteor it automatically creates a new project directory. User's have requested that we allow them to specify the new project or select an existing project to load.

### CODE REVIEW

File: src/odemis/gui/cont/tabs/cryo_chamber_tab.py [hunk 1]
```diff
@@ -910,6 +924,37 @@ def _sample_heater_to_va(self):
...
         Tab.Show(self, show=show)
 
+        # the first time the tab is shown, ask the user to create or load a project
+        if self.IsShown() and not self._is_initial_project_ready:
+
+            # stay in the loop until the user successfully creates or loads a project
+            while True:
+
+                # Ask the user to create or load a project
+                box = wx.MessageDialog(
```
Comment (nandishjpatel) [inline src/odemis/gui/cont/tabs/cryo_chamber_tab.py:934]: Maybe you can handle this dialog and the subsequent dialogues (steps -> pages (buttons) in Wizard) using https://docs.wxpython.org/wx.adv.Wizard.html

Just a suggestion!

Comment (patrickcleeve2) [reply src/odemis/gui/cont/tabs/cryo_chamber_tab.py:934]: Interesting, didn't know about this! Probably not for this PR, but might be worth re-doing the create/load project workflow with this

File: src/odemis/gui/cont/tabs/cryo_chamber_tab.py [hunk 2]
```diff
@@ -910,6 +924,37 @@ def _sample_heater_to_va(self):
...
+                    self.main_frame,
+                    message="Create or Load a Meteor Project?",
```
Comment (nandishjpatel) [inline src/odemis/gui/cont/tabs/cryo_chamber_tab.py:936]:
```suggestion
                    message="Create or Load a Meteor Project",
```

File: src/odemis/gui/cont/tabs/cryo_chamber_tab.py [hunk 3]
```diff
@@ -370,15 +382,16 @@ def _reset_project_data(self):
 
     def _load_project_data(self, evt: wx.Event):
```
Comment (nandishjpatel) [inline src/odemis/gui/cont/tabs/cryo_chamber_tab.py:383]:
```suggestion
    def _load_project_data(self, evt: wx.Event) -> bool:
```

?

File: src/odemis/gui/cont/tabs/cryo_chamber_tab.py [hunk 4]
```diff
@@ -260,15 +261,16 @@ def _on_change_project_folder(self, evt):
         Shows a dialog to change the path and name of the project directory.
         returns nothing, but updates .conf and project path text control
         """
-        # TODO: do not warn if there is no data (eg, at init)
```
Comment (nandishjpatel) [inline src/odemis/gui/cont/tabs/cryo_chamber_tab.py:263]: `def _on_change_project_folder(self, evt) -> Optional[bool]:`

File: src/odemis/gui/cont/tabs/cryo_chamber_tab.py [hunk 5]
```diff
@@ -353,6 +359,12 @@ def _create_new_dir(self):
 
         self._change_project_conf(np)
 
+    def _create_project(self):
```
Comment (nandishjpatel) [inline src/odemis/gui/cont/tabs/cryo_chamber_tab.py:362]:
```suggestion
    def _create_project(self) -> Optional[bool]:
```

File: src/odemis/gui/cont/tabs/cryo_chamber_tab.py [hunk 6]
```diff
@@ -910,6 +924,37 @@ def _sample_heater_to_va(self):
     def Show(self, show=True):
         Tab.Show(self, show=show)
 
+        # the first time the tab is shown, ask the user to create or load a project
+        if self.IsShown() and not self._is_initial_project_ready:
+
+            # stay in the loop until the user successfully creates or loads a project
+            while True:
```
Comment (nandishjpatel) [inline src/odemis/gui/cont/tabs/cryo_chamber_tab.py:931]: Just wondering if you can do away with this while statement. If not, its fine!

**Review by nandishjpatel — APPROVED** [review]

File: src/odemis/gui/cont/tabs/cryo_chamber_tab.py (hunk 3 above)
Comment (pieleric) [reply src/odemis/gui/cont/tabs/cryo_chamber_tab.py:383]: I think it's supposed to be `bool`. Please add a docstring to explain what False & True mean.

Then, every place with just `return` should be `return False`.

File: src/odemis/gui/cont/tabs/cryo_chamber_tab.py [hunk 7]
```diff
@@ -90,8 +90,9 @@ def __init__(self, name, button, panel, main_frame, main_data):
         self.txt_projectpath = self.panel.txt_projectpath
         self.btn_load_project.Hide()
 
-        # Create new project directory on starting the GUI
-        self._create_new_dir()
+        # indicates whether a project has been initially created or loaded
+        self._is_initial_project_ready: bool = False
```
Comment (pieleric) [inline src/odemis/gui/cont/tabs/cryo_chamber_tab.py:94]: I don't really get what this means. Does this mean that the data is explicitly stored in a project selected by the user (in contrast to the data being stored in the default new project folder created at init)?
Could you clarify this in the comment (or the variable name).

**Review by pieleric — CHANGES_REQUESTED** [review]

File: src/odemis/gui/cont/tabs/cryo_chamber_tab.py [hunk 8]
```diff
@@ -910,6 +924,37 @@ def _sample_heater_to_va(self):
...
+                    style=wx.YES_NO | wx.ICON_QUESTION | wx.CENTER,
+                )
+
+                box.SetYesNoLabels("&Load Project", "&Create Project",)
+                ans = box.ShowModal()  # Waits for the window to be closed
+
+                if ans == wx.ID_NO:
+                    logging.debug("creating project")
+                    ret = self._create_project()
+                if ans == wx.ID_YES:
```
Comment (pieleric) [inline src/odemis/gui/cont/tabs/cryo_chamber_tab.py:947]: `elif`

File: src/odemis/gui/cont/tabs/cryo_chamber_tab.py (hunk 6 above)
Comment (patrickcleeve2) [reply src/odemis/gui/cont/tabs/cryo_chamber_tab.py:931]: No, we want the user to stay here until they either create a new project or load an existing one. So if they fail to do so, they get put back to the start of the flow until they do so correctly.

File: src/odemis/gui/cont/tabs/cryo_chamber_tab.py (hunk 7 above)
Comment (patrickcleeve2) [reply src/odemis/gui/cont/tabs/cryo_chamber_tab.py:94]: It's a flag to indicate whether the user has initially created a new project or loaded an existing one. The rest of the tab is blocked until they succesffuly do that.

**Review by tmoerkerken — APPROVED** [review]

**Review by pieleric — APPROVED** [review]

**Pull request merged**